bgpd/bgp_nht.c

```diff
@@ -320,10 +321,12 @@ int bgp_find_or_add_nexthop(struct bgp *bgp_route, struct bgp *bgp_nexthop,
 		/* updates NHT pi list reference */
 		path_nh_map(pi, bnc, true);
 
+		bpi_ultimate = bgp_get_imported_bpi_ultimate(pi);
 		if (CHECK_FLAG(bnc->flags, BGP_NEXTHOP_VALID) && bnc->metric)
-			(bgp_path_info_extra_get(pi))->igpmetric = bnc->metric;
-		else if (pi->extra)
-			pi->extra->igpmetric = 0;
+			(bgp_path_info_extra_get(bpi_ultimate))->igpmetric =
+				bnc->metric;
+		else if (bpi_ultimate->extra)
+			bpi_ultimate->extra->igpmetric = 0;
```

Comment:
Can we use the same getter here as well? bgp_path_info_extra_get

Reply:
That's the idea to keep the code reusable and changing/moving attributes anywhere in the future without a bunch of changes. Getter/Setter helpers are very handy in such situations.

Reply:
bgp_path_info_extra_get is overkill. It checks the presence of extra and creates it if not there.